### PR TITLE
fix v1 profile not exists v2 exists case

### PR DESCRIPTION
**Pull request by KutluOzel-b**
Signed-off-by: KutluOzel-b <kutlu.ozel@broadcom.com>

## Proposed changes

Fix registerCustomProfileType for extenders throwing error while v1 profile not in use but v2 in use. 

Checking if v1 or v2 is used if so then registers profile.

## Release Notes

Milestone:

https://github.com/zowe/vscode-extension-for-zowe/issues/1870

Changelog:

## Types of changes

- [ x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

- [ x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [x ] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ ] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ x] There is coverage for the code that I have added
- [ ] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 1]
```diff
@@ -123,8 +123,7 @@ export class ProfilesCache {
      * @param {string} type Type of Profile, defaults to "zosmf" if nothing passed.
      *
      * @returns {IProfileLoaded[]}
-     */
-    public getProfiles(type = "zosmf"): zowe.imperative.IProfileLoaded[] {
+     */ public getProfiles(type = "zosmf"): zowe.imperative.IProfileLoaded[] {
```
Comment (roman-kupriyanov) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:126]: Probably, unnecessary change here.

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 2]
```diff
@@ -136,14 +135,17 @@ export class ProfilesCache {
...
      */
-    public registerCustomProfilesType(profileTypeName: string): void {
-        const exists = fs.existsSync(path.posix.join(`${os.homedir()}/.zowe/profiles/${profileTypeName}`));
-        if (!exists) {
-            throw new Error(
-                `Zowe Explorer Profiles Cache error: Tried to register a custom profile type named: ${profileTypeName} that does not yet exist. Extenders must call "zoweExplorerApi.getExplorerExtenderApi().initForZowe()" first.`
-            );
-        }
-        this.allExternalTypes.add(profileTypeName);
+    public registerCustomProfilesType(profileTypeName: string) {
```
Comment (roman-kupriyanov) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:138]: According to the `eslint` rules the function return type is required. I would leave the `void` type in place.

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 3]
```diff
@@ -136,14 +135,17 @@ export class ProfilesCache {
...
+        const existsV1Profile = fs.existsSync(path.posix.join(`${os.homedir()}/.zowe/profiles/${profileTypeName}`));
+        this.getProfileInfo().then((res) => {
```
Comment (roman-kupriyanov) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:140]: This call requires a full `then..catch` block according to the `eslint` rules again. So I would suggest to add simple:
```
.then(...)
.catch((error) => { throw error; });
```
to fix the build.

Comment (roman-kupriyanov) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:140]: An alternative solution might be to convert `registerCustomProfilesType` function into `async` function to fully use `await` operator. But since this is a public API, you may want to introduce a new `asyncRegisterCustomProfilesType` call and deprecate the existing one.

This approach would be preferable, because otherwise we are loosing error throwing from `registerCustomProfilesType` function itself (they will stuck in callbacks).

**Review by roman-kupriyanov — COMMENTED** [review]
Hi @KutluOzel-b! Thank you very much for this change! 😃 

I tested it with our E4E extension which uses Zowe Extenders API very heavily and I can confirm that it fixed our issues 🥳 

Though I have some suggestions/improvements for the fixes, please consider them before releasing the fix.

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts (hunk 3 above)
Comment (KutluOzel-b) [reply packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:140]: Thanks Roman your inputs much appreciated :)

File: packages/zowe-explorer-api/src/profiles/ProfilesCache.ts [hunk 4]
```diff
@@ -131,7 +131,7 @@ export class ProfilesCache {
     /**
      * Used for extenders to register with Zowe Explorer that do not need their
      * profile type in the existing MVS, USS, and JES
-     *
+     * @deprecated Please use asyncRegisterCustomProfilesType
```
Comment (roman-kupriyanov) [inline packages/zowe-explorer-api/src/profiles/ProfilesCache.ts:134]:
```ts
    /**
     * @deprecated Please use {@link asyncRegisterCustomProfilesType}.
     * 
     * Used for extenders to register with Zowe Explorer that do not need their
     * profile type in the existing MVS, USS, and JES
     *
     * @param {string} profileTypeName Type of Profile, defaults to "zosmf" if nothing passed.
     *
     * @returns {void}
     */
```

Minor: a proposal to use a link for the new function and move `@deprecated` annotation to the top of the JSDoc.

**Review by roman-kupriyanov — APPROVED** [review]
Thanks for the updates @KutluOzel-b 👍 

I recheck once more with our extension, it still works and the errors are handled properly. Thank you again 😃 

I have a small suggestion for the deprecation message, please consider to apply.

**Review by roman-kupriyanov — DISMISSED** [review]

**Comment by JillieBeanSim** [comment]
Looks like we have a DCO issue on this branch, the fix is stated here https://github.com/zowe/vscode-extension-for-zowe/pull/1885/checks?check_run_id=7520732371

**Comment by KutluOzel-b** [comment]
Rebase conflict, closing this pr and opening a new one.
